docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,364 @@
+
+> Note: Dependent ordering among tasks will be addresed later which may enable concurrent execution of tasks.
+
+#### Deletion
+
+`Etcd-druid` will ensure that it garbage collects the custom resources and any other k8s resources created to realise the task as per `.spec.ttlSecondsAfterFinished` if defined in the [spec](#spec) or it assumes a default expiry time.
+
+### Use Cases
+
+#### Recovery from permanent quorum loss
+
+Currently identification and recovery from permament quorum loss is done manaually. The current proposal keeps the identification as manual, requiring an human operator to intervene and confirm that there is indeed a permanent quorum loss with no auto healing possibility. Once that is established then the next step to recover the cluster is now proposed to be automated. Recovery today is a [multi-step process](https://github.com/gardener/etcd-druid/blob/master/docs/operations/recovery-from-permanent-quorum-loss-in-etcd-cluster.md) and needs to be done carefully. Automation would ensure that we eliminate any errors from an operator.
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should be a quorum loss in a multi-member etcd cluster. For a single-member etcd cluster, invoking this task is unnecessary as the restoration of the single member is automatically handled by the backup-restore process.
+* There should not already be a permanent-quorum-loss-recovery-task running for the same etcd cluster.
+
+#### Trigger on-demand snapshot compaction
+
+`etcd-druid` provides a configurable [etcd-events-threshold](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/02-snapshot-compaction.md#druid-flags). If and when this threshold is breached then only a [snapshot compaction](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/02-snapshot-compaction.md) is triggered for etcd cluster. However, there are scenarios where an ad-hoc snapshot compaction is required.
+
+##### Possible scenarios
+
+* Full snapshot taken via compaction job is better as compared to full snapshot taken directly from etcd cluster as full snapshot taken via snapshot compaction is taken after [defragmentation](https://etcd.io/docs/v3.2/op-guide/maintenance/#defragmentation) which makes it smaller (more compacted) in size.
+* Full snapshot triggered on etcd cluster might cause load on running etcd cluster but full snapshot taken by compaction job is developed separately with zero effect on the running cluster.
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should not be a `on-demand snapshot-compaction` task already running for the same etcd cluster.
```

Comment:
Again a thought, will it be more clean, if this operation creates an compaction job on demand instead.  
In this case if there is any subsequent compaction job which was about to get triggered with delta's reaching the threshold, then it will not run as the compaction already has a logic to check for any running jobs before initiating a new one.

Reply:
>if this operation creates an compaction job on demand instead.

Did not understand. All of these use cases are on-demand/out-of-band tasks.

Reply:
>There should not be a `on-demand snapshot-compaction` task already running for the same etcd cluster. 
I don't see the current solution can provide a deterministic behaviour here unless you ensure you run on-demand snapshot compaction also as a job like the regular compaction-snapshot job. 

Case in hand.
compaction delta threshold: 100
t(-1): deltas: 99
t0: no compaction job running; (ondemand snapshot is triggered)
t0: deltas:100 --> compaction job is triggered
t1: can t0 happen?

Reply:
>  In this case if there is any subsequent compaction job which was about to get triggered with delta's reaching the threshold, then it will not run as the compaction already has a logic to check for any running jobs before initiating a new one.

yes, that's correct but why you want to run `on demand snapshot-compaction` when `snapshot compaction` is already about to get triggered ... main motive of `on-demand snapshot-compaction` is operator don't want to wait for `event-threshold` to reach. Anyway, for the scenario you have describe, one of the compaction job won't run, Isn't this a correct behaviour ?

>  unless you ensure you run on-demand snapshot compaction also as a job like the regular compaction-snapshot job.

here, we are not changing any behaviour of `snapshot compaction` feature. so, yes `on-demand snapshot compaction` will also be run as `compaction job`. 

> t0: no compaction job running; (ondemand snapshot is triggered)

did you mean `ondemand compaction` ? If yes then only 1 `snapshot compaction` job will be triggered.

Reply:
Let me rephrase the ask may be that's simpler to explain. 
> There should not be a `on-demand snapshot-compaction` task already running for the same etcd cluster.

How are you ensuring this given that we have 2 actors now:
1. Regular Snapshot Compaction Job 
2. OnDemand Compaction 

Why do you say it can never lead to race condition.

Reply:
>yes, that's correct but why you want to run on demand snapshot-compaction when snapshot compaction is already about to get triggered .

But the operator will not know about this or will he? 
You should consider the fact that this can be used by anyone who has the access and might not be aware that there is a job running or the job might not be aware that there is a compaction executed onDemand. 

I've got to understand that we should avoid 2 actors and have deterministic behavior from @unmarshall, so I was wondering if this guiding principle is voilated here. 

If you think that's not the case, then its fine we can close this remark.

Reply:
as you already mentioned in the comment https://github.com/gardener/etcd-druid/pull/757#discussion_r1550948932 that compaction feature already has a capability to detect old job before initiating a new one then we are fine as we are not changing any behaviour of snapshot compaction feature.
Here, out-of-band task controller will act just as an orchestrator to trigger the `snapshot compaction feature`. IMO, this edge case won't cause any un-deterministic behaviour as `snapshot compaction feature` checking this.

https://github.com/gardener/etcd-druid/blob/637da10c12b8df72e46532e04a0fd4768ced60d9/controllers/compaction/reconciler.go#L123-L124